chain/src/chain.rs

```diff
@@ -367,32 +382,41 @@ impl Chain {
 	/// This is only ever used during sync and is based on sync_head.
 	/// We update header_head here if our total work increases.
 	pub fn sync_block_headers(&self, headers: &[BlockHeader], opts: Options) -> Result<(), Error> {
+		let mut sync_pmmr = self.sync_pmmr.write();
+		let mut header_pmmr = self.header_pmmr.write();
 		let mut txhashset = self.txhashset.write();
-		let batch = self.store.batch()?;
```

Comment:
We create two separate block contexts and these both take ownership of the batch, so we can't reuse the same one.

We need to rework the block context to take a mutable ref to the batch. Pretty sure this is possible but the code gets tricky because of the lifetimes involved.
Maybe tackle this in a separate PR.

To you initial question - yes, we lock the txhashset and the header MMRs so we can safely create and commit multiple batches within this block of code.